### PR TITLE
LPS-136720 Implement options to disable and configure link toolbar

### PR DESCRIPTION
**Pull request by markocikos**
JIRA: https://issues.liferay.com/browse/LPS-136720

Steps to test:
1. deploy `frontend-editor-ckeditor-sample-web`
2. change `toolbarLink` in `config` to modify or disable link toolbar.

The gif below contains a sample of customization. In it, I added `Bold` button and removed `LinkBrowse` button. To get it, set
```js
toolbarLink: 'Bold,LinkRemoveUnlink,LinkTarget,LinkInput,LinkConfirm',
```

This is in contrast to default
```js
toolbarLink: 'LinkRemove,LinkTarget,LinkInput,LinkConfirm,LinkBrowse',
```

To disable link toolbar, delete `toolbarLink`.

Notes:
- Any standard ckeditor button, like `Bold`, can be added. Some might not work, depending on context.
- Somebody can make a custom plugin with a toolbar item created with `editor.ui.addBalloonToolbarButton`, add it to config, and it will work. There might be some creative hacking needed if they want to interact with other parts of toolbar, like we are doing with default toolbar items. 
- A side-commit to the main bulk of this PR is https://github.com/liferay-frontend/liferay-portal/commit/f21ae4592659e10e309841ad9e94fbcf1466e842, a fix to selection edge case, that popped up during testing.

After PR 
after

### CODE REVIEW

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/linktoolbar/plugin.js [hunk 1]
```diff
@@ -66,197 +106,156 @@
...
+							selection = editor.getSelection();
 
-					const range = ranges[0];
+							const selectedText = selection.getSelectedText();
 
-					const bookmark = range.createBookmark();
+							const ranges = selection.getRanges();
 
-					linkElement = new CKEDITOR.dom.element('a');
+							const range = ranges[0];
```
Comment (diegonvs) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/linktoolbar/plugin.js:204]: @markocikos, I needed to debug and see the result of `ranges` to understand what the first position was. Can We go with:

```js
const [rangeElement] = selection.getRanges();
```

this API is very weird 😂 

Screen Shot 2021-08-12 at 20 29 52

Comment (markocikos) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/linktoolbar/plugin.js:204]: @diegonvs API kinda makes sense, if you have in mind that the goal is to define position of selected area. They do this by breaking selected area into rectangles, that can be defined by `x`, `y`, `height` and `width`. In example below, `ranges` will have three elements in array:

screen

So, to get to the start of selection, you go to start of the first rectangle. There is an alternative, probably cleaner way to to do this, using `selection.getStartElement()`, but this has some limitations and quirky behavior, see even in this PR the `startElement.getAscendant` bit.

In this PR, I didn't change the behavior of `LinkConfirm`. I only renamed it from `okButton` and moved it to `_initToolbarItems`, along with other custom toolbar items.

As for using 
```js
const [firstRange] = selection.getRanges();
```
yeah, that works nicely! :+1:  I'm not used to using destructuring in this way, but I think I will start to moving forward.

Comment (markocikos) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/linktoolbar/plugin.js:204]: Nevermind me, I was thinking of `range.getClientRect()` :sweat_smile:  That gives the useful array of rectangles. I am not sure what's the case with multiple `range`s. Maybe some parts of ckeditor can have multiselection?

**Comment by markocikos** [comment]
@diegonvs I force-pushed. Changes:
1. SF commit, https://github.com/liferay-frontend/liferay-portal/pull/1350/commits/a633372aa81ec48d754e6f1521efdd0b00daa04e, based on https://github.com/liferay-frontend/liferay-portal/pull/1350#discussion_r688148856
2. I reworked the side-commit with selection exclusion, it didn't correctly fix the issue. See gif below, how the text toolbar pops up behind link toolbar. The solution is in https://github.com/liferay-frontend/liferay-portal/pull/1350/commits/dcca4d0c95fe8fe3c299d25021d71e23b935dd87

Before PR:
before

After PR:
after

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/linktoolbar/plugin.js (hunk 1 above)
Comment (diegonvs) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/linktoolbar/plugin.js]: I was taking a look in what means bookmark and I saw nothing unless circular pointings betweens createBooksmarks is used to create a bookmark to use in moveBookmark but I don't know what it is. 

Seeing source code now I understand what bookmark is, From: https://github.com/ckeditor/ckeditor4/blob/4e64f672190a14d3c38aebd94ae4fb10c79e4545/core/dom/range.js#L708

** a bookmark object, which can be later used to restore the range by using the {@link #moveToBookmark} function.

😂

**Comment by markocikos** [comment]
On hold until https://github.com/brianchandotcom/liferay-portal/pull/105601 is merged.

**Comment by markocikos** [comment]
Force-pushed. Changes:
- Fixed conflict, by moving `toolbarLink` to newly added `DefaultBalloonEditorConfiguration.js`
- Removed commit that was fixing multiple toolbars bug, splitting it off into a separate task: https://github.com/liferay/liferay-ckeditor/pull/189

No other changes in code. As this was reviewed, except for trivial conflict fix, we can forward.

**Comment by markocikos** [comment]
ci:forward

**Comment by diegonvs** [comment]
ci:forward

**Comment by markocikos** [comment]
@john-co Please take a look at `WYSIWYGUsecase#CanDisplayContentRightToLeft `. CI is failing on in in this PR and in https://github.com/liferay-frontend/liferay-portal/pull/1369

**Comment by markocikos** [comment]
ci:forward

**Pull request closed without merging**